### DISCUSSION TITLE
构建特定类型的DNS请求来测试结果的可信度

### DISCUSSION THREAD

**Discussion by PussAzuki** [IrineSistiana/mosdns#367] [Ideas]
转自：https://www.v2ex.com/t/857393

省流：国内IP请求任意国外IP(或反之)，均存在“被污染”的域名被抢答的行为，能不能利用这点？

**Comment by PussAzuki** [comment]
最简单的思路：
准备与国内/国外的DNS上游若干(都至少一个)，然后新插件(?或许是)做如下行为：每次收到A/AAAA请求时，同时向国外(此处IP应用户自定义；若客户端实际IP为国外，则向国内)发送一次TXT请求，然后用户设置根据是否返回TXT结果来决定转发到国内还是国外DNS又或者从并发中保留结果

**Comment by PussAzuki** [comment]
理论上和 _prefer_ipv4/_prefer_ipv6 差不多的逻辑，但不知道实践起来有没有意外

**Comment by IrineSistiana** [comment]
不行，实践起来没用稳定性保证。

这种探测非常不稳定，不能拿来做实时判断。mosdns 这种dns 程序需要稳定的实时判断。理论上确实和 _prefer_ipv4/_prefer_ipv6 差不多的逻辑，但后者的有服务器提供商保证探测请求的稳定性。

个人觉得这个思路非常适合本地生成 “被污染列表” 用。生成后作为静态表丢给 mosdns 用。

**Reply by PussAzuki** [reply]
现在或许可以先写个功能出来测试下？
我看着多请求一次TXT记录到9.9.9.9，如果是被污染的域名，那么额外延迟只有10+ms（因为是国内就被干掉了）。
也就是说构建一个TXT请求，限制为短时间的超时时间，超时就当无污染，然后直接先送到国内dns解析，再国外dns解析，反之则直接国外dns解析。

**Reply by PussAzuki** [reply]
试了下存在 *.googleusercontent.com 这种没有被污染但是被屏蔽的网站...确实不完全靠谱

只能抓到那些dns污染...

**Comment by PussAzuki** [comment]
国内用户请求国外IP/国外用户请求国内IP -> "被污染"/"不被污染" ,看起来只有两种情况，稳定性应该有保障的吧

我的思路是 第一次可能被污染(上游超时/服务失败->先返回一个短TTL的结果优先保证时间)，但之后决不会被污染(如果只抢答A/AAAA，那么不是便宜我们了吗)

那么该怎么处理“被污染”域名列表的匹配规则？应该使用domain还是full？
